File: Data Structures & Algorithms/count-paths/submission-2.java
class Solution {
    public int uniquePaths(int m, int n) {
        int[][] dp = new int[m][n];
        for(int i = 0; i < m; i++)
            Arrays.fill(dp[i], -1);
        return helper(0, 0, m, n, dp);
    }

    public int helper(int r, int c, int rows, int cols, int[][] dp){
        if (r == rows || c == cols)
            return 0;
        if (r == rows - 1 && c == cols - 1)
            return 1;
        if(dp[r][c] != -1)
            return dp[r][c];
        dp[r][c] = helper(r + 1, c, rows, cols, dp) + helper(r, c + 1, rows, cols, dp);
        return dp[r][c];
    }
}

File: Data Structures & Algorithms/rotting-fruit/submission-2.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        rows, cols = len(grid), len(grid[0])
        directions = [(0, 1), (0, -1), (1, 0), (-1, 0)]

        fresh, q = 0, deque()
        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == 1:
                    fresh += 1
                elif grid[r][c] == 2:
                    q.append((r, c))

        time = 0
        while q and fresh:
            for _ in range(len(q)):
                r, c = q.popleft()
                for dr, dc in directions:
                    nr, nc = r + dr, c + dc
                    if 0<=nr<rows and 0 <=nc<cols and grid[nr][nc] == 1:
                        fresh -= 1
                        grid[nr][nc] = 2
                        q.append((nr, nc))
            time += 1
        return time if not fresh else -1
        